contracts/vortex/CarbonVortex.sol

```diff
@@ -138,109 +345,508 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
+        uint128 targetAmount
+    ) external payable nonReentrant validToken(token) greaterThanZero(targetAmount) {
+        uint128 sourceAmount;
+        if (token == _targetToken) {
+            sourceAmount = _sellFinalTargetForTarget(targetAmount);
+        } else {
+            sourceAmount = _sellTargetTokenForToken(token, targetAmount);
+        }
+        emit TokenTraded({ caller: msg.sender, token: token, sourceAmount: sourceAmount, targetAmount: targetAmount });
+    }
+
+    function _sellTargetTokenForToken(Token token, uint128 targetAmount) private returns (uint128) {
+        uint128 sourceAmount = expectedTradeInput(token, targetAmount);
+        // revert if trade requires 0 target token
+        if (sourceAmount == 0) {
+            revert InvalidTrade();
+        }
+        // check enough target token (if target token is native) has been sent for the trade
+        if (_targetToken == NATIVE_TOKEN && msg.value < sourceAmount) {
+            revert InsufficientNativeTokenSent();
+        } else {
+            // if target token is not native, transfer the tokens to the contract
+            _targetToken.safeTransferFrom(msg.sender, address(this), sourceAmount);
+        }
+        // transfer the tokens to caller
+        token.safeTransfer(msg.sender, targetAmount);
+
+        // if the target token is native, refund any excess native token to caller
+        if (_targetToken == NATIVE_TOKEN && msg.value > sourceAmount) {
+            payable(msg.sender).sendValue(msg.value - sourceAmount);
+        }
+
+        // if no final target token is defined, transfer the target token to `transferAddress`
+        if (Token.unwrap(_finalTargetToken) == address(0)) {
+            // safe due to nonreenrant modifier (forwards all available gas if token is native)
+            _targetToken.unsafeTransfer(_transferAddress, sourceAmount);
+            // increment total collected in `transferAddress`
+            _totalCollected += sourceAmount;
+        }
+
+        // if remaining balance is below the min token sale amount, reset the auction
+        if (_amountAvailableForTrading(token) < _minTokenSaleAmounts[token]) {
+            _resetTrading(token, 0);
+        }
+
+        return sourceAmount;
+    }
+
+    function _sellFinalTargetForTarget(uint128 targetAmount) private returns (uint128) {
+        uint128 sourceAmount = expectedTradeInput(_targetToken, targetAmount);
+        // revert if trade requires 0 finalTarget tokens
+        if (sourceAmount == 0) {
+            revert InvalidTrade();
+        }
+
+        // check enough final target token (if final target token is native) has been sent for the trade
+        if (_finalTargetToken == NATIVE_TOKEN) {
+            if (msg.value < sourceAmount) {
+                revert InsufficientNativeTokenSent();
+            } else {
+                // transfer the tokens to the _transferAddress
+                payable(_transferAddress).sendValue(sourceAmount);
```

Comment:
What about returning any leftover to the caller, i.e., if `msg.value > sourceAmount`?

Reply:
we do it 10 lines below